docs/source/library-user-guide/working-with-exprs.md

```diff
@@ -19,4 +19,205 @@
+
+- Simplifying `Expr`s to make them easier to evaluate
+- Optimizing `Expr`s to make them faster to evaluate
+- Converting `Expr`s to other forms, e.g. converting a `BinaryExpr` to a `CastExpr`
+
+In our example, we'll use rewriting to update our `add_one` UDF, to be rewritten as a `BinaryExpr` with a `Literal` of 1. We're effectively inlining the UDF.
+
+### Rewriting with `transform`
+
+To implement the inlining, we'll need to write a function that takes an `Expr` and returns a `Result<Expr>`. If the expression is _not_ to be rewritten `Transformed::No` is used to wrap the original `Expr`. If the expression _is_ to be rewritten, `Transformed::Yes` is used to wrap the new `Expr`.
+
+```rust
+fn rewrite_add_one(expr: Expr) -> Result<Expr> {
+    expr.transform(&|expr| {
+        Ok(match expr {
+            Expr::ScalarUDF(scalar_fun) => {
+                // rewrite the expression if the function is "add_one", otherwise return the original expression
+                if scalar_fun.fun.name == "add_one" {
+                    let input_arg = scalar_fun.args[0].clone();
+
+                    let new_expression = BinaryExpr::new(
+                        Box::new(input_arg),
+                        datafusion::logical_expr::Operator::Plus,
+                        Box::new(Expr::Literal(datafusion::scalar::ScalarValue::Int64(Some(
+                            1,
+                        )))),
+                    );
+
+                    Transformed::Yes(Expr::BinaryExpr(new_expression))
+                } else {
+                    // a scalar function that is not "add_one" is not rewritten
+                    Transformed::No(Expr::ScalarUDF(scalar_fun))
+                }
+            }
+            _ => Transformed::No(expr),  // not a scalar function, so not rewritten
+        })
+    })
+}
+```
+
+### Creating an `OptimizerRule`
+
+In DataFusion, an `OptimizerRule` is a trait that supports rewriting `Expr`s. It follows DataFusion's general mantra of trait implementations to drive behavior.
+
+We'll call our rule `AddOneInliner` and implement the `OptimizerRule` trait. The `OptimizerRule` trait has two methods:
+
+- `name` - returns the name of the rule
+- `try_optimize` - takes a `LogicalPlan` and returns an `Option<LogicalPlan>`. If the rule is able to optimize the plan, it returns `Some(LogicalPlan)` with the optimized plan. If the rule is not able to optimize the plan, it returns `None`.
+
+```rust
+struct AddOneInliner {}
+
+impl OptimizerRule for AddOneInliner {
+    fn name(&self) -> &str {
+        "add_one_inliner"
+    }
+
+    fn try_optimize(
+        &self,
+        plan: &LogicalPlan,
+        config: &dyn OptimizerConfig,
+    ) -> Result<Option<LogicalPlan>> {
+        // recurse down and optimize children first
+        let optimized_plan = utils::optimize_children(self, plan, config)?;
+
+        match optimized_plan {
```

Comment:
I think you can use `plan.expressions()` and then `from_plan`: https://docs.rs/datafusion/latest/datafusion/logical_expr/utils/fn.from_plan.html to rewrite expressions anywhere they appear.

It took me a while to find "from_plan" -- I will propose moving it to LogicalPlan as a method.

Reply:
Thanks for this, it seems to clean things up nicely, though please lmk if I'm not using correctly. Also, I updated the code to use the new method proposed in https://github.com/apache/arrow-datafusion/pull/7396... if that doesn't go in I can update this back.